File: Episode02-Add_Mobs/Pygame/mob.py
import pygame, random
import shared

class Mob():
	def __init__(self, w:int, h:int, colour:tuple) -> None:
		self.rect = pygame.Rect(0, 0, w, h)		# create a rectangle at 0,0 measuring w x h	
		self.colour = colour					# set colour
		self.reset()

	def reset(self) -> None:
		# allowing for the width of the sprite itself
		self.rect.x = random.randrange(0, shared.WIDTH - self.rect.width)
		# start off the top of the screen by random amount
		self.rect.y = random.randrange(-150, -100)
		# set speed to a random amount, so some will move faster than others
		self.speedy = random.randrange(50, 600)
		self.speedx = random.randrange(-200, 200)		
			
	def update(self, dt:float) -> None:
		self.rect.y	+= round(self.speedy * dt)
		self.rect.x += round(self.speedx * dt)
		if	self.rect.top > shared.HEIGHT + 10 or self.rect.left < -25 or self.rect.right > shared.WIDTH + 20:
			self.reset()
			
	def draw(self) -> None:
		pygame.draw.rect(shared.screen, self.colour, self.rect)

File: Episode04-Add_Graphics/Love2D/mob.lua
local Class = require("lib.Class")
local Mob = Class:derive("Mob")
local Rectangle = require "lib.Rectangle"
local Shared = require "shared"

function Mob:new(sprite)
	--[[ class constructor now takes a sprite parameter]]
	self.sprite = sprite
	self.rect = Rectangle(0, 0, sprite:getWidth(), sprite:getHeight())
	self:setProperties()
end

function Mob:setProperties()
	-- set speed to a random amount, so some will move faster than others
	self.rect.x = math.random(0, Shared.WIDTH - self.sprite:getWidth())	-- make the mob appear on screen randomly across it's width
	self.rect.y = math.random(-150, -100)			-- start off the top of the screen by random amount
	self.speedY = math.random(50, 600)
	self.speedX = math.random(-100, 100)
	if Shared.debug then
		self.speedY = math.random(10, 100)
		self.speedX = math.random(-20, 20)
	end
end

function Mob:getRect()
	return self.rect
end

function Mob:update(dt)
	self.rect:updatePosition(self.speedX * dt, self.speedY * dt)
	-- check if Mob has gone off bottom or sides of screen
	if self.rect.top > Shared.HEIGHT or self.rect.right < 0 or self.rect.left > Shared.WIDTH then
		self:setProperties() -- reset to position above top of the screen
	end
end

function Mob:draw()
	love.graphics.draw(self.sprite, self.rect.x, self.rect.y)
	if Shared.debug then
		love.graphics.setColor(Shared.RED)
		love.graphics.rectangle("line", self.rect.x, self.rect.y, self.rect.w, self.rect.h)
		love.graphics.setColor(Shared.WHITE)
	end
end

return Mob